Add clarifications and details to integration-tests README [skip ci] (#4694)
NVIDIA/spark-rapids: integration_tests/README.md

  [here](https://github.com/pyenv/pyenv#installation)
- Verify that `pyenv` is set correctly
  
  ```shell script
  which pyenv  
  ```

- Using `pyenv` to set Python installation
  - To check versions to be installed (will return a long list)
  
    ```shell script
    ls ~/.pyenv/versions/
    ```

  - To install a specific version from the available list
  
    ```shell script
    pyenv install 3.X.Y
    ```

  - To check available versions locally

    ```shell script
    ls ~/.pyenv/versions/
    ```

  - To set python environment to one of the installed versions

    ```shell script
    pyenv global 3.X.Y
    ```

For full details on `pyenv` and instructions, see [pyenv github page](https://github.com/pyenv/pyenv).

#### Installing specific version of Maven

All package managers like `brew` and `apt` offers maven. However, it may lag behind some
versions. In that case, you can install the latest binary from the [Maven download page](https://maven.apache.org/download.cgi).
For manual installation, you need to setup your environment:

  ```shell script
  export M2_HOME=PATH_TO_MAVEN_ROOT_DIRECTOTY
  export M2=${M2_HOME}/bin
  export PATH=$M2:$PATH
  ```

> </details>

Reviewer: This looks like a closing tag with no matching `<details>` tag.

verify that the plugin is doing the right thing in as many cases as possible.

There are two sets of tests. The PySpark tests are described on this page. The scala tests are
described [here](../tests/README.md).

> ## Setting Environment

Reviewer: ```suggestion
## Setting Up the Environment
```

The tests are based off of `pyspark` and `pytest` running on Python 3. There really are
only a small number of Python dependencies that you need to install for the tests. The
dependencies also only need to be on the driver.  You can install them on all nodes
in the cluster but it is not required.

### Prerequisities

The build requires `OpenJDK 8`, `maven`, and `python`.

> skip to the next section if you have already installed them.

Reviewer: ```suggestion
Skip to the next section if you have already installed them.
```

<details><summary><b>Java Environment</b></summary>
<p>

It is recommended to use `alternatives` to manage multiple java versions.
Then you can simply set `JAVA_HOME` to JDK directory:

  ```shell script
  JAVA_HOME=$(readlink -nf $(which java) | xargs dirname | xargs dirname | xargs dirname)
  ```

</p>
</details>

<details><summary><b>Installing python using pyenv</b></summary>

> <p>

Reviewer: Curious, do we need the `<b>` and `<p>` tags?  For the former we would normally use `*` to mark bold text, and the latter is normally done with simple newlines.  One of the advantages of Markdown is it doesn't have a lot of HTML tag pollution in the text, and it would be nice to minimize it here as well for those reading the README in a terminal window rather than rendered on Github.  I see we're using Markdown inside here, so I'm wondering if some of these HTML tags are strictly necessary.

Author: The html tags would make the header collapsible.
Previously, `<p>` had to be there to be parse the feature.
I removed the collapsible headers to avoid breaking the coding style.

It is recommended that you use `pyenv` to manage Python installations.

- First, make sure to install all the required dependencies listed
  [here](https://github.com/pyenv/pyenv/wiki#suggested-build-environment).
- Follow instructions to use the right method of installation described
  [here](https://github.com/pyenv/pyenv#installation)
- Verify that `pyenv` is set correctly
  
  ```shell script
  which pyenv  
  ```

- Using `pyenv` to set Python installation
  - To check versions to be installed (will return a long list)
  
    ```shell script
    ls ~/.pyenv/versions/
    ```

  - To install a specific version from the available list
  
    ```shell script
    pyenv install 3.X.Y
    ```

  - To check available versions locally

    ```shell script
    ls ~/.pyenv/versions/
    ```

  - To set python environment to one of the installed versions

    ```shell script
    pyenv global 3.X.Y
    ```

For full details on `pyenv` and instructions, see [pyenv github page](https://github.com/pyenv/pyenv).
</p>
</details>

> <details><summary><b>Installing specific version of maven</b></summary>

Reviewer: ```suggestion
<details><summary><b>Installing specific version of Maven</b></summary>
```

<p>

> All package managers like `brew` and `apt` offers maven. However, it may lag behind some

Reviewer: ```suggestion
All package managers like `brew` and `apt` offers Maven. However, it may lag behind some
```

- sre_yield
  : Provides a set of APIs to generate string data from a regular expression.
- pandas
  : A fast, powerful, flexible and easy to use open source data analysis and manipulation
  tool and is only needed when testing integration with pandas.
- pyarrow
  : Provides a Python API for functionality provided by the Arrow C++ libraries, along with
  tools for Arrow integration and interoperability with pandas, NumPy, and other software in
  the Python ecosystem. This is used to test improved transfer performance to pandas based user
  defined functions.
- pytest-xdist
  : A plugin that extends pytest with new test execution modes, the most used being distributing
  tests across multiple CPUs to speed up test execution
- findspark
  : Adds pyspark to sys.path at runtime

You can install all the dependencies using `pip` by running the following command:

  ```shell script
  pip install pytest \
              sre_yield \
              pandas \
              pyarrow \
              pytest-xdist \
              findspark
  ```

### Installing Spark

You need to install spark-3.x and set `SPARK_HOME/bin` to your `$PATH`.

> To build a distribution run the following command. Full instructions can be found [here](https://spark.apache.org/docs/latest/building-spark.html).

Reviewer: There doesn't seem to be a "following command" to build a distribution here, it just goes onto another paragraph without a command?

> To run the plugin against the latest spark, you will need to buid a new version from source:

Reviewer: ```suggestion
To run the plugin against a snapshot version of Spark, you will need to build Spark from source:
```

  ```shell script

>   ## to build a distribution from branch is branch-3.x.y

Reviewer: Spark does not use a patch version for branches, it's just `branch-x.y` (e.g.: Spark 3.2.2-SNAPSHOT source is on `branch-3.2`).
```suggestion
  ## The source for Spark snapshot version X.Y.Z is on branch-x.y
  ```

  SPARK_BRANCH=branch-3.x.y
  SPARK_VER=3xy
  
  git clone -b ${SPARK_BRANCH} https://github.com/apache/spark.git spark-src-${SPARK_VER}
  cd spark-src-${SPARK_VER}
  ## build a distribution with hive support
  ./dev/make-distribution.sh --name spark-for-rapids-${SPARK_VER} --tgz -Pkubernetes -Phive
  ## this will generate a single tgz file that you would need to untar it to $SPARK_INSTALLS_DIR
  ```

Set the environment:

  ```shell script
  SPARK_HOME=$SPARK_INSTALLS_DIR/spark-${SPARK_VER}
  export SPARK_HOME
  export PATH=${SPARK_HOME}/bin:$PATH
  ```

### Building The Plugin

> Make sure that you compile the plugin against the same version of Spark that it is going to run with.

Reviewer: This should probably just link to the contributor's guide which covers how to build the plugin.  Putting a copy of instructions here is likely to decay into incorrect instructions over time.
